flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/ApplyFlintSparkSkippingIndex.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.flint.spark.skipping
+
+import org.opensearch.flint.spark.FlintSpark
+import org.opensearch.flint.spark.skipping.FlintSparkSkippingIndex.{getSkippingIndexName, FILE_PATH_COLUMN, SKIPPING_INDEX_TYPE}
+
+import org.apache.spark.sql.Column
+import org.apache.spark.sql.catalyst.expressions.{And, Predicate}
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, LogicalRelation}
+import org.apache.spark.sql.flint.FlintDataSourceV2.FLINT_DATASOURCE
+
+/**
+ * Flint Spark skipping index apply rule that rewrites applicable query's filtering condition and
+ * table scan operator to leverage additional skipping data structure and accelerate query by
+ * reducing data scanned significantly.
+ *
+ * @param flint
+ *   Flint Spark API
+ */
+class ApplyFlintSparkSkippingIndex(flint: FlintSpark) extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case filter @ Filter( // TODO: abstract pattern match logic for different table support
+          condition: Predicate,
+          relation @ LogicalRelation(
+            baseRelation @ HadoopFsRelation(location, _, _, _, _, _),
+            _,
+            Some(table),
+            false)) =>
+      // Exit if plan is already rewritten with skipping index
+      if (location.isInstanceOf[FlintSparkSkippingFileIndex]) {
+        return plan
+      }
+
+      val indexName = getSkippingIndexName(table.identifier.table) // TODO: database name
+      val index = flint.describeIndex(indexName)
+      if (index.exists(_.kind == SKIPPING_INDEX_TYPE)) {
+        val skippingIndex = index.get.asInstanceOf[FlintSparkSkippingIndex]
+        val indexPred = rewriteToIndexPredicate(skippingIndex, condition)
+
+        /*
+         * Replace original file index with Flint skipping file index:
+         *  Filter(a=b)
+         *  |- LogicalRelation(A)
+         *     |- HadoopFsRelation
+         *        |- FileIndex <== replaced with FlintSkippingFileIndex
+         */
+        if (indexPred.isDefined) {
+          val selectedFiles = selectFilesByIndex(skippingIndex, indexPred.get)
+          val fileIndex = new FlintSparkSkippingFileIndex(location, selectedFiles)
```

Comment:
could selectedFiles be lazy evaluated?

Reply:
Sure, this may block query optimizer. Will pass DataFrame to file index class instead. Thanks!

Reply:
Addressed in https://github.com/opensearch-project/sql/pull/1690/commits/db4e44bf8b13c6611529646e370712abb78606ac.